bootstrap: vendor crates required by opt-dist to collect profiles

weihanglo opened the pull request:
These are the default package set required by opt-dist to correctly work,
hence for people wanting to build a production grade of rustc in a
sandboxed / air-gapped environment, these need to be vendored.

The size of `rustc-src-nightly.tar.xz` before and after this change:

* Before: 298M
* After: 323M (+8%)

Size change might or might not be a concern.
See the previous discussion: https://github.com/rust-lang/rust/pull/125166#issuecomment-2113626468

Previous efforts on making:

* https://github.com/rust-lang/rust/pull/125125
* https://github.com/rust-lang/rust/pull/125166

---

Note that extra works still need to be done to make it fully vendored.

* The current pinned rustc-perf uses `tempfile::Tempdir` as the working
  directory when collecting profiles from some of these packages.
  This "tmp" working directory usage make it impossible for Cargo to pick
  up the correct vendor sources setting in `.cargo/config.toml` bundled
  in the rustc-src tarball. [^1]
* opt-dist verifies the final built rustc against a subset of rustc test
  suite. However it rolls out its own `config.toml` without setting
  `vendor = true`, and that results in `./vendor/` directory removed.
  [^2]

[^1]: https://github.com/rust-lang/rustc-perf/blob/4f313add609f43e928e98132358e8426ed3969ae/collector/src/compile/benchmark/mod.rs#L164-L173
[^2]: https://github.com/rust-lang/rust/blob/606afbb617a2949a4e35c4b0258ff94c980b9451/src/tools/opt-dist/src/tests.rs#L62-L77

weihanglo committed:
feat: vendor crates required by opt-dist to collect profiles

These are the default package set required by opt-dist to correctly work,
hence for people wanting to build a production grade of rustc in a
sandboxed / air-gapped environment, these need to be vendored.

The size of `rustc-src-nightly.tar.xz` before and after this change:

* Before: 298M
* After: 323M (+8%)

These crates are the default set of packages required by opt-dist
to correctly work, hence for people wanting to build a production grade
of rustc in an sandboxed / air-gapped environment, these need to be vendored.

The size of `rustc-src-nightly.tar.xz` before and after this change:

* Before: 298M
* After: 323M (+8%)

Size change might or might not be a concern.
See the previous discussion: https://github.com/rust-lang/rust/pull/125166#issuecomment-2113626468

Previous efforts on making:

* https://github.com/rust-lang/rust/pull/125125
* https://github.com/rust-lang/rust/pull/125166

---

Note that extra works still need to be done to make it fully vendored.

* The current pinned rustc-perf uses `tempfile::Tempdir` as the working
  directory when collecting profiles from some of these packages.
  This "tmp" working directory usage make it impossible for Cargo to pick
  up the correct vendor sources setting in `.cargo/config.toml` bundled
  in the rustc-src tarball. [^1]
* opt-dist verifies the final built rustc against a subset of rustc test
  suite. However it rolls out its own `config.toml` without setting
  `vendor = true`, and that results in `./vendor/` directory removed.
  [^2]

[^1]: https://github.com/rust-lang/rustc-perf/blob/4f313add609f43e928e98132358e8426ed3969ae/collector/src/compile/benchmark/mod.rs#L164-L173
[^2]: https://github.com/rust-lang/rust/blob/606afbb617a2949a4e35c4b0258ff94c980b9451/src/tools/opt-dist/src/tests.rs#L62-L77

weihanglo committed:
docs: add README.md for build_helper

weihanglo commented:
r? @Mark-Simulacrum as you reviewed the previous two PRs.
cc @Kobzol as you have been working hard in this area :)

Kobzol commented:
r? @Kobzol

This makes sense, and it is nicer than packaging everything from `rustc-perf`, which is quite large.

@bors r+ rollup

bors commented:
:pushpin: Commit 51a87fcc3f50e79d71046ee8181b99c9130a3913 has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol 51a87fcc3f50e79d71046ee8181b99c9130a3913 -->
<!-- homu: {"type":"Approved","sha":"51a87fcc3f50e79d71046ee8181b99c9130a3913","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r- 
I guess this failed here https://github.com/rust-lang/rust/pull/125491#issuecomment-2129831457

weihanglo committed:
tidy: add license exceptions for rustc-perf

`ring` is included because it is an optional dependency of `hyper`,
which is a training data in rustc-pref for optimized build.
The license of it is generally `ISC AND MIT AND OpenSSL`,
though the `package.license` field is not set.

See https://github.com/briansmith/ring/issues/902

`git+https://github.com/rust-lang/team` git source is from
`rust_team_data`, which is used by `site` in src/tools/rustc-perf.
This doesn't really a part of the compiler,
so doesn't really affect the bootstrapping process.

See https://github.com/rust-lang/rustc-perf/pull/1914.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9232782084/job/25404414231) [(plain)](https://github.com/rust-lang/rust/commit/0b1b705d3afbaf622501c5ade8b123c22c6b2c7a/checks/25404414231/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:a5ac7e51b41094c92402da3b24376905380afc29)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.454   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.466 Collecting boolean-py==4.0
#13 0.469   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.483 Collecting chardet==5.1.0
---
#13 3.488 Building wheels for collected packages: reuse
#13 3.489   Building wheel for reuse (pyproject.toml): started
#13 3.816   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.817   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.818   Stored in directory: /tmp/pip-ephem-wheel-cache-7dknseb_/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.820 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.843   Attempting uninstall: setuptools
#13 3.843     Found existing installation: setuptools 59.6.0
#13 3.845     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.845     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.845     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.532 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.532 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 5.056 Collecting virtualenv
#13 5.110   Downloading virtualenv-20.26.2-py3-none-any.whl (3.9 MB)
#13 5.170      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 68.3 MB/s eta 0:00:00
#13 5.207 Collecting distlib<1,>=0.3.7
#13 5.213   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.250 Collecting filelock<4,>=3.12.2
#13 5.253   Downloading filelock-3.14.0-py3-none-any.whl (12 kB)
#13 5.253   Downloading filelock-3.14.0-py3-none-any.whl (12 kB)
#13 5.284 Collecting platformdirs<5,>=3.9.1
#13 5.288   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.376 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.541 Successfully installed distlib-0.3.8 filelock-3.14.0 platformdirs-4.2.2 virtualenv-20.26.2
#13 DONE 5.6s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      208832 kB
DirectMap2M:     8179712 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/21e6de7eb64c09102de3f100420a09edc1a2a8d7/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-21e6de7eb64c09102de3f100420a09edc1a2a8d7-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
    Finished `release` profile [optimized] target(s) in 25.62s
##[endgroup]
fmt check
tidy check
tidy error: invalid source: "git+https://github.com/rust-lang/measureme?tag=9.2.0#9f51cde2e5dd3ef0392f0f6a7201f4946502ef41"
tidy error: invalid source: "git+https://github.com/rust-lang/measureme?branch=master#8f1b8321b28e49139a86f07ce44dc88ef61b38d5"
tidy error: invalid source: "git+https://github.com/rust-lang/measureme?branch=master#8f1b8321b28e49139a86f07ce44dc88ef61b38d5"
tidy error: invalid source: "git+https://github.com/rust-lang/measureme?tag=9.2.0#9f51cde2e5dd3ef0392f0f6a7201f4946502ef41"
tidy error: invalid source: "git+https://github.com/rust-lang/measureme?branch=master#8f1b8321b28e49139a86f07ce44dc88ef61b38d5"
tidy error: invalid source: "git+https://github.com/rust-lang/team#80ae2e8c7a509fcbfaefaedebac51b3bd4dfc4b8"
tidy error: dependency `git+https://github.com/rust-lang/measureme?branch=master#decodeme@10.1.1` does not define a license expression
tidy error: dependency `registry+https://github.com/rust-lang/crates.io-index#ring@0.16.20` does not define a license expression
tidy error: dependency `git+https://github.com/rust-lang/team#rust_team_data@1.0.0` does not define a license expression
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.0)
Collecting black==23.3.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 7))
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 43.3 MB/s eta 0:00:00
Collecting click==8.1.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 34))
  Downloading click-8.1.3-py3-none-any.whl (96 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.6/96.6 kB 33.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.6/96.6 kB 33.0 MB/s eta 0:00:00
Collecting importlib-metadata==6.7.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 38))
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
Collecting mypy-extensions==1.0.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 42))
  Downloading mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)
Collecting packaging==23.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 46))
  Downloading packaging-23.1-py3-none-any.whl (48 kB)
Collecting pathspec==0.11.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 50))
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
Collecting platformdirs==3.6.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 54))
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
Collecting ruff==0.0.272 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 58))
  Downloading ruff-0.0.272-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.9 MB)
Collecting tomli==2.0.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 77))
  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)
Collecting typed-ast==1.5.4 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 81))
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 877.7/877.7 kB 110.4 MB/s eta 0:00:00
Collecting typing-extensions==4.6.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 107))
  Downloading typing_extensions-4.6.3-py3-none-any.whl (31 kB)
Collecting zipp==3.15.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 114))
  Downloading zipp-3.15.0-py3-none-any.whl (6.8 kB)
Installing collected packages: zipp, typing-extensions, typed-ast, tomli, ruff, platformdirs, pathspec, packaging, mypy-extensions, click, importlib-metadata, black
Successfully installed black-23.3.0 click-8.1.3 importlib-metadata-6.7.0 mypy-extensions-1.0.0 packaging-23.1 pathspec-0.11.1 platformdirs-3.6.0 ruff-0.0.272 tomli-2.0.1 typed-ast-1.5.4 typing-extensions-4.6.3 zipp-3.15.0
some tidy checks failed
Build completed unsuccessfully in 0:01:02
  local time: Sat May 25 04:12:15 UTC 2024
  network time: Sat, 25 May 2024 04:12:15 GMT
```

</details>

weihanglo commented:
License exceptions added. Some issues I am not sure how to resolve.

* Git sources are not allowed, though they are vendorable. Is it okay to add them to [the exception source list](https://github.com/rust-lang/rust/blob/606afbb617a2949a4e35c4b0258ff94c980b9451/src/tools/tidy/src/extdeps.rs#L7)?
* `rust-lang/team` and `rust-lang/measureme` have no license. I guess it's fine just adding one for them?

```
tidy error: invalid source: "git+https://github.com/rust-lang/measureme?tag=9.2.0#9f51cde2e5dd3ef0392f0f6a7201f4946502ef41"
tidy error: invalid source: "git+https://github.com/rust-lang/measureme?branch=master#8f1b8321b28e49139a86f07ce44dc88ef61b38d5"
tidy error: invalid source: "git+https://github.com/rust-lang/measureme?branch=master#8f1b8321b28e49139a86f07ce44dc88ef61b38d5"
tidy error: invalid source: "git+https://github.com/rust-lang/measureme?tag=9.2.0#9f51cde2e5dd3ef0392f0f6a7201f4946502ef41"
tidy error: invalid source: "git+https://github.com/rust-lang/measureme?branch=master#8f1b8321b28e49139a86f07ce44dc88ef61b38d5"
tidy error: invalid source: "git+https://github.com/rust-lang/team#80ae2e8c7a509fcbfaefaedebac51b3bd4dfc4b8"
tidy error: dependency `git+https://github.com/rust-lang/measureme?branch=master#decodeme@10.1.1` does not define a license expression
tidy error: dependency `registry+https://github.com/rust-lang/crates.io-index#ring@0.16.20` does not define a license expression
tidy error: dependency `git+https://github.com/rust-lang/team#rust_team_data@1.0.0` does not define a license expression
```

bjorn3 commented:
> Git sources are not allowed, though they are vendorable. Is it okay to add them to [the exception source list](https://github.com/rust-lang/rust/blob/606afbb617a2949a4e35c4b0258ff94c980b9451/src/tools/tidy/src/extdeps.rs#L7)?

I think they should only be allowed for the crates that are vendored in this PR. In general builds should not require git dependencies as otherwise non-vendored builds could break if the git repo is gone or removes the commit in question.

Mark-Simulacrum commented:
> rust-lang/team and rust-lang/measureme have no license. I guess it's fine just adding one for them?

Yeah, we should propagate upstream and get approvals from relevant contributors for each of those repos (shouldn't be many people).

michaelwoerister commented:
Where are these old measureme versions coming from?

Kobzol commented:
I'm not sure if the list contains everything needed currently though. Does it contain just the dependencies of `rustc-perf`, or also the dependencies of the compile-time benchmarks that we use for the PGO workflow?

weihanglo commented on src/tools/tidy/src/deps.rs:
```diff
@@ -67,6 +67,7 @@ pub(crate) const WORKSPACES: &[(&str, ExceptionList, Option<(&[&str], &[&str])>)
     //("src/tools/miri/test-cargo-miri", &[], None), // FIXME uncomment once all deps are vendored
     //("src/tools/miri/test_dependencies", &[], None), // FIXME uncomment once all deps are vendored
     ("src/tools/rust-analyzer", EXCEPTIONS_RUST_ANALYZER, None),
+    ("src/tools/rustc-perf", EXCEPTIONS_RUSTC_PERF, None),
```
> davidtwco and I talked about this today and we think it makes sense to add the `EXCEPTIONS_RUSTC_PERF` list and we're ok with the current set of crates/licenses in it. It would be best if we can avoid growing the list further, just to make maintenance easier but this seems like the correct approach to us. 

> I'm not sure if the list contains everything needed currently though. Does it contain just the dependencies of `rustc-perf`, or also the dependencies of the compile-time benchmarks that we use for the PGO workflow?

@wesleywiser @Kobzol 

It currently only includes dependencies of rustc-perf, as those training data crates are not part of the rustc-perf workspace (also the reason we manually `cargo vendor` them). Should we license-check those training data crates as well?

(moved to a thread for following the discussion easier)

Kobzol replied:
I *think* that we should check the licenses of everything that we vendor, but I'm not 100% sure.

weihanglo replied:
> check the licenses of everything that we vendor

Including the web frontend code? 🤯 
https://github.com/rust-lang/rustc-perf/blob/master/site/frontend/package.json

Kobzol replied:
My assumption was that we should check licences of everything that we vendor, although I'm not sure if we currently do that. If that is needed, it would probably be the simplest if we just deleted the frontend code before packaging it in the source tarball.

weihanglo commented:
The license-check issue is resolved. Everything is ready :)

Thanks @Kobzol and @michaelwoerister for the speedy review on measureme/rustc-perf side.

Note that two exceptions are added:

`ring` is included because it is an optional dependency of `hyer`,
which is a training data in rust-pref for optimized build.
The license of it is generally `ISC AND MIT AND OpenSSL`,
though the `package.license` field is not set.

See https://github.com/briansmith/ring/issues/902

`git+[https://github.com/rust-lang/team`](https://github.com/rust-lang/team%60) git source is from
`rust_team_data`, which is used by `site` in src/tools/rustc-perf.
This doesn't really a part of the compiler,
so doesn't really affect the bootstrapping process.

See https://github.com/rust-lang/rustc-perf/pull/1914.

Kobzol commented on src/tools/tidy/src/deps.rs:
```diff
@@ -178,6 +195,20 @@ const EXCEPTIONS_UEFI_QEMU_TEST: ExceptionList = &[
+/// These dependencies have non-standard licenses but are genenrally permitted.
+const EXCEPTIONS_NON_STANDARD_LICENSE_DEPS: &[&str] = &[
+    // `ring` is included because it is an optional dependency of `hyer`,
+    // which is a training data in rust-pref for optimized build.
```
```suggestion
    // which is a training data in rustc-perf for optimized build.
```

weihanglo replied:
Nice catch! Fixed

Kobzol commented on src/tools/tidy/src/deps.rs:
```diff
@@ -178,6 +195,20 @@ const EXCEPTIONS_UEFI_QEMU_TEST: ExceptionList = &[
+
+/// These dependencies have non-standard licenses but are genenrally permitted.
+const EXCEPTIONS_NON_STANDARD_LICENSE_DEPS: &[&str] = &[
+    // `ring` is included because it is an optional dependency of `hyer`,
```
```suggestion
    // `ring` is included because it is an optional dependency of `hyper`,
```

Kobzol reviewed (commented):
Few typos.

weihanglo commented:
@Kobzol, this is ready for another round of review

@wesleywiser, due to `ring` using `package.license-file`, I've added a `EXCEPTIONS_NON_STANDARD_LICENSE_DEPS` check to bypass that. It is actually `ISC AND MIT AND OpenSSL` and is used as optimized build training data. It is supposed to be fine I believe?

Kobzol commented:
I'm not sure about the licensing part, but the rest looks good to me. Have you tried to do a vendored build of `opt-dist`, to see if it works end-to-end? We probably won't have any testing for this.

(Big thanks for pushing all these changes through, btw! It's great to provide the optimized workflow for more people that build their own compiler).

weihanglo commented:
@Kobzol yes. I ran `src/ci/docker/run.sh dist-x86_64-linux` with some modifications and it works. No crate was downloaded during the build. Verified by checking the existence of `$CARGO_HOME/registry,git}`. Without `--benchmark-cargo-config` provided crates will be downloaded.


<details><summary>

Click to see the diff (basically call `opt-dist local` with `linux-ci` args)

</summary>
<p>

```patch
diff --git a/src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile b/src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile
index 4e1aae98221..f1911fdea96 100644
--- a/src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile
+++ b/src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile
@@ -66,6 +66,7 @@ ENV PGO_HOST=x86_64-unknown-linux-gnu
 ENV HOSTS=x86_64-unknown-linux-gnu
 
 ENV RUST_CONFIGURE_ARGS \
+      --enable-vendor \
       --enable-full-tools \
       --enable-sanitizers \
       --enable-profiler \
@@ -80,10 +81,22 @@ ENV RUST_CONFIGURE_ARGS \
       --set rust.lto=thin \
       --set rust.codegen-units=1
 
-ENV SCRIPT python3 ../x.py build --set rust.debug=true opt-dist && \
-    ./build/$HOSTS/stage0-tools-bin/opt-dist linux-ci -- python3 ../x.py dist \
+ENV SCRIPT mkdir -p /tmp/tmp-multistage/ && \
+  ln -s /checkout/.cargo /tmp/tmp-multistage/.cargo && \
+  ln -s /checkout/vendor /tmp/tmp-multistage/vendor && \
+  python3 ../x.py build --set rust.debug=true opt-dist && \
+    ./build/$HOSTS/stage0-tools-bin/opt-dist local \
+    --target-triple x86_64-unknown-linux-gnu \
+    --checkout-dir /checkout \
+    --artifact-dir /tmp/tmp-multistage/opt-artifacts \
+    --llvm-dir /rustroot \
+    --llvm-shared \
+    --benchmark-cargo-config /checkout/.cargo/config.toml \
+    --use-bolt \
+    -- python3 ../x.py dist \
     --host $HOSTS --target $HOSTS \
     --include-default-paths \
+    --skip-tests tests/ui/process/nofile-limit.rs \
     build-manifest bootstrap
 ENV CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_LINKER=clang
 
diff --git a/src/tools/opt-dist/src/main.rs b/src/tools/opt-dist/src/main.rs
index e4271a6e2dd..c141ff72c9a 100644
--- a/src/tools/opt-dist/src/main.rs
+++ b/src/tools/opt-dist/src/main.rs
@@ -132,7 +132,7 @@ fn create_environment(args: Args) -> anyhow::Result<(Environment, Vec<String>)>
                 .checkout_dir(checkout_dir.clone())
                 .host_llvm_dir(llvm_dir)
                 .artifact_dir(artifact_dir)
-                .build_dir(checkout_dir)
+                .build_dir(checkout_dir.join("obj"))
                 .prebuilt_rustc_perf(rustc_perf_checkout_dir)
                 .shared_llvm(llvm_shared)
                 .use_bolt(use_bolt)

```

</p>
</details>

Kobzol commented:
Awesome :)

weihanglo commented:
> > rust-lang/team and rust-lang/measureme have no license. I guess it's fine just adding one for them?
> 
> Yeah, we should propagate upstream and get approvals from relevant contributors@m for each of those repos (shouldn't be many people).

@Mark-Simulacrum I think the license and git source issues have been resolved. Would you mind reviewing this? We're vendoring `rustc-perf` in 1.80.0, so it might better for this to go hand-in-hand with that before new nightly branches off.

Mark-Simulacrum commented:
Nightly probably already branched, but I'll take a look on the weekend and maybe nominate for beta backport depending on details.

Mark-Simulacrum commented on src/tools/build_helper/src/metrics.rs:
```diff
@@ -1,5 +1,28 @@
 use serde_derive::{Deserialize, Serialize};
 
+/// The default set of crates for opt-dist to collect LLVM profiles.
```
Not sure metrics makes sense for this? I'd put it in `lib.rs` I guess...

Mark-Simulacrum commented on src/tools/tidy/src/deps.rs:
```diff
@@ -178,6 +195,20 @@ const EXCEPTIONS_UEFI_QEMU_TEST: ExceptionList = &[
+/// These dependencies have non-standard licenses but are genenrally permitted.
+const EXCEPTIONS_NON_STANDARD_LICENSE_DEPS: &[&str] = &[
+    // `ring` is included because it is an optional dependency of `hyper`,
+    // which is a training data in rustc-pref for optimized build.
```
```suggestion
    // which is a training data in rustc-perf for optimized build.
```

Mark-Simulacrum commented:
r=me with nits fixed

> maybe nominate for beta backport depending on details.

I think given the changes here I'm not currently inclined to backport it.

weihanglo commented:
nits are fixed, and the CI is all green now. Thanks for the review people.


@bors r=Mark-Simulacrum

bors commented:
:pushpin: Commit 377870386f99ee811e8945fade761a3ea8944db1 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 377870386f99ee811e8945fade761a3ea8944db1 -->
<!-- homu: {"type":"Approved","sha":"377870386f99ee811e8945fade761a3ea8944db1","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

michaelwoerister commented:
🎉
